spec/02-integration/05-proxy/16-origins_spec.lua

```diff
@@ -0,0 +1,61 @@
+local helpers = require "spec.helpers"
+
+describe("origins config option", function()
```

Comment:
Please add tests for the proper matching of incoming (non-normalized) addresses to the normalized ones in `origins`.

Reply:
@James Callahan with my current understanding of the code it will simply fail if you configure a service with an upstream url `http://192.168.001.001/somewhere` due to a lack of normalizing the ipv4.
Also be aware that plugins might change the upstream properties, so the normalization must be done in each request. Just updating the service entity is not enough.

Reply:
fyi have a look at this example code: https://github.com/Tieske/ncache.lua/blob/master/examples/ipcache.lua#L241-L258

Reply:
IPv4 addresses are not allowed to contain leading zeros: that should be a syntax error according to RFC 3986. If we don't check that on input then that is where the fix needs to be.

Reply:
That does not matter. Since the utility functions are now inconsistent. And those functions serve a bigger purpose than just this change.

And even then the test cases should be added to show it is validated in the right places.

And finally the function is expensive and should not be on a hot code path imo.